### PR TITLE
Implement variables relevant for group stage to playoff transition

### PR DESCRIPTION
**Pull request by Malaber**

### CODE REVIEW

File: app/models/tournament.rb [hunk 1]
```diff
@@ -24,4 +26,11 @@ def generate_code
       break if errors['code'].blank?
     end
   end
+
+  def playoff_teams_amount_is_positive_power_of_two
+    return if (Utils.po2?(playoff_teams_amount) && playoff_teams_amount.positive?) || playoff_teams_amount.zero?
+
+    errors.add(:playoff_teams_amount_isnt_positive_power_of_two,
```
Comment (Thor77) [inline app/models/tournament.rb]: you should add the error to a field, not a random symbol

File: spec/interactors/add_group_stage_to_tournament_interactor_spec.rb [hunk 2]
```diff
@@ -32,6 +40,16 @@
         test = empty_tournament_context.tournament.stages.first
         expect(test).to eq(@group_stage)
       end
+
+      it 'sets default for instant_finalists_amount' do
+        test = empty_tournament_context.tournament.instant_finalists_amount
```
Comment (Thor77) [inline spec/interactors/add_group_stage_to_tournament_interactor_spec.rb]: Please at least don't overwrite [built-in functions](https://devdocs.io/ruby~2.5/kernel#method-i-test) (if you really think this var assignment is necssary here)

File: spec/interactors/add_group_stage_to_tournament_interactor_spec.rb [hunk 3]
```diff
@@ -32,6 +40,16 @@
...
+        expect(test).to eq(@tournament_service_defaults.first)
+      end
+
+      it 'sets default for intermediate_round_participants_amount' do
+        test = empty_tournament_context.tournament.intermediate_round_participants_amount
```
Comment (Thor77) [inline spec/interactors/add_group_stage_to_tournament_interactor_spec.rb]: Please at least don't overwrite [built-in functions](https://devdocs.io/ruby~2.5/kernel#method-i-test) (if you really think this var assignment is necssary here)

File: spec/services/tournament_service_spec.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+RSpec.describe TournamentService do
+  describe '#calculate_default_amount_of_teams_advancing', focus: true do
```
Comment (Thor77) [inline spec/services/tournament_service_spec.rb]: please remove `focus: true`, we don't need those filters on our master branch

File: spec/interactors/add_group_stage_to_tournament_interactor_spec.rb [hunk 5]
```diff
@@ -32,6 +40,16 @@
         test = empty_tournament_context.tournament.stages.first
```
Comment (Thor77) [inline spec/interactors/add_group_stage_to_tournament_interactor_spec.rb]: i know i should've catched this earlier, but please rename the var here, too

File: spec/controllers/tournaments_controller_spec.rb [hunk 6]
```diff
@@ -234,10 +257,17 @@
...
+          it 'renders an unprocessable entity response' do
+            expect(response).to have_http_status(:unprocessable_entity)
+          end
         end
+
+        before do
+          put :create, params: context
+        end
+
+        context ''
```
Comment (Thor77) [inline spec/controllers/tournaments_controller_spec.rb]: ?

File: spec/controllers/tournaments_controller_spec.rb [hunk 7]
```diff
@@ -234,10 +257,17 @@
       end
 
       context 'with invalid parameters' do
-        it 'renders an unprocessable entity response' do
-          put :create, params: { teams: [1, 2, 3] }
-          expect(response).to have_http_status(:unprocessable_entity)
+        shared_examples_for 'renders unprocessable entity response' do
```
Comment (Thor77) [inline spec/controllers/tournaments_controller_spec.rb]: please remove unused examples

File: spec/interactors/add_playoffs_to_tournament_interactor_spec.rb [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # frozen_string_literal: true
 
-RSpec.describe AddPlayoffsToTournament do
+RSpec.describe AddPlayoffsToTournament, type: :interactor do
```
Comment (Thor77) [inline spec/interactors/add_playoffs_to_tournament_interactor_spec.rb:3]: what's the purpose of this?

File: spec/controllers/tournaments_controller_spec.rb [hunk 9]
```diff
@@ -282,6 +312,103 @@
...
+              end
+                .to_not(change { @filled_tournament.playoff_teams_amount })
+            end
+
+            it 'doesn\'t change instant_finalists_amount' do
+              expect do
+                put :update, params: { id: @filled_tournament.to_param }.merge(playoff_teams_amount: 8)
+                @filled_tournament.reload
+              end
+                .to_not(change { @filled_tournament.instant_finalists_amount })
```
Comment (Thor77) [inline spec/controllers/tournaments_controller_spec.rb:375]: this probably belongs into the line above

File: spec/controllers/tournaments_controller_spec.rb (hunk 7 above)
Comment (Thor77) [reply spec/controllers/tournaments_controller_spec.rb]: (or revert your change here, it doesn't look like that case is tested anymore)

File: spec/controllers/tournaments_controller_spec.rb [hunk 10]
```diff
@@ -194,6 +194,29 @@
             expect(@group_stage_tournament.playoff_teams_amount)
               .to eq(create_group_tournament_data[:playoff_teams_amount])
           end
+
+          context 'playoff_teams_amount unacceptable' do
+            it 'is not a power of two' do
+              post :create, params: create_group_tournament_data.merge(playoff_teams_amount: 18)
+              expect(response).to have_http_status(:unprocessable_entity)
+              expect(deserialize_response(response).values.first.first)
```
Comment (Thor77) [inline spec/controllers/tournaments_controller_spec.rb]: i would prefer referring to the error by keys instead of this, but if that's not possible i'm fine with this

File: spec/controllers/tournaments_controller_spec.rb [hunk 11]
```diff
@@ -194,6 +194,29 @@
             expect(@group_stage_tournament.playoff_teams_amount)
               .to eq(create_group_tournament_data[:playoff_teams_amount])
           end
+
+          context 'playoff_teams_amount unacceptable' do
+            it 'is not a power of two' do
```
Comment (Thor77) [inline spec/controllers/tournaments_controller_spec.rb]: this doesn't really read like a valid it-block.. they should read like actions and not input-conditions..

**Review by Thor77 — CHANGES_REQUESTED** [review]

File: spec/controllers/tournaments_controller_spec.rb (hunk 9 above)
Comment (Malaber) [reply spec/controllers/tournaments_controller_spec.rb:375]: I put it down there because I found it hard to read, was pretty cluttered

File: spec/controllers/tournaments_controller_spec.rb (hunk 11 above)
Comment (Malaber) [reply spec/controllers/tournaments_controller_spec.rb]: so should I put a context around it saying context 'isn't a power of two' and the it part then saying 'it fails' ?

File: spec/interactors/add_playoffs_to_tournament_interactor_spec.rb (hunk 8 above)
Comment (Malaber) [reply spec/interactors/add_playoffs_to_tournament_interactor_spec.rb:3]: That is what they autogenerate with, so I thought I would put it there for the sake of completion

File: spec/services/tournament_service_spec.rb (hunk 4 above)
Comment (Malaber) [reply spec/services/tournament_service_spec.rb]: done

File: spec/interactors/add_group_stage_to_tournament_interactor_spec.rb (hunk 2 above)
Comment (Malaber) [reply spec/interactors/add_group_stage_to_tournament_interactor_spec.rb]: done

File: spec/interactors/add_group_stage_to_tournament_interactor_spec.rb (hunk 3 above)
Comment (Malaber) [reply spec/interactors/add_group_stage_to_tournament_interactor_spec.rb]: done

File: spec/interactors/add_group_stage_to_tournament_interactor_spec.rb (hunk 5 above)
Comment (Malaber) [reply spec/interactors/add_group_stage_to_tournament_interactor_spec.rb]: done

File: spec/controllers/tournaments_controller_spec.rb (hunk 6 above)
Comment (Malaber) [reply spec/controllers/tournaments_controller_spec.rb]: removed

File: spec/controllers/tournaments_controller_spec.rb (hunk 7 above)
Comment (Malaber) [reply spec/controllers/tournaments_controller_spec.rb]: did so

File: spec/controllers/tournaments_controller_spec.rb (hunk 10 above)
Comment (Malaber) [reply spec/controllers/tournaments_controller_spec.rb]: I don't know what you mean, it might be possible

File: app/models/tournament.rb (hunk 1 above)
Comment (Malaber) [reply app/models/tournament.rb]: What do you mean? I don't know what that is for..

Comment (Thor77) [reply app/models/tournament.rb]: See [this page](https://guides.rubyonrails.org/active_record_validations.html#custom-methods) for an explanation and example. `errors.add` is used to add errors to fields (fields as in form fields transmitted to the API)

File: spec/controllers/tournaments_controller_spec.rb (hunk 11 above)
Comment (Thor77) [reply spec/controllers/tournaments_controller_spec.rb]: That would be the "correct" way to do it, but in this case that would cause three contexts with just one it-block in them.. so it's probably fine to leave it like this or think of another way to structure this specific test

File: spec/controllers/tournaments_controller_spec.rb (hunk 9 above)
Comment (Thor77) [reply spec/controllers/tournaments_controller_spec.rb:375]: Looks weird to me having it hang around there, but if you think it improves readability, I'm fine with that

File: app/models/tournament.rb (hunk 1 above)
Comment (Malaber) [reply app/models/tournament.rb]: done i guess

File: spec/controllers/tournaments_controller_spec.rb (hunk 9 above)
Comment (Malaber) [reply spec/controllers/tournaments_controller_spec.rb:375]: it does look weird, yes, but I would leave it like that if it is okay for you.. wont kill me if I change it to be in one line :D

File: spec/controllers/tournaments_controller_spec.rb (hunk 11 above)
Comment (Malaber) [reply spec/controllers/tournaments_controller_spec.rb]: is that better? 

https://github.com/turniere/turniere-backend/pull/55/commits/22040f7600f6d1c58936e98268b1da58910253e6

File: spec/controllers/tournaments_controller_spec.rb [hunk 12]
```diff
@@ -282,6 +311,103 @@
           expect(response).to have_http_status(:ok)
           expect(response.content_type).to eq('application/json')
         end
+
+        context 'any variable relevant for group stage to playoff transition changed' do
+          before do
```
Comment (Thor77) [inline spec/controllers/tournaments_controller_spec.rb]: ?

File: spec/controllers/tournaments_controller_spec.rb [hunk 13]
```diff
@@ -233,13 +269,6 @@
         end
       end
 
-      context 'with invalid parameters' do
```
Comment (Thor77) [inline spec/controllers/tournaments_controller_spec.rb]: why did you remove this?

**Review by Thor77 — CHANGES_REQUESTED** [review]

File: spec/controllers/tournaments_controller_spec.rb (hunk 10 above)
Comment (Thor77) [reply spec/controllers/tournaments_controller_spec.rb]:
```ruby
expect(deserialize_response(response)[:playoff_teams_amount].first)
```

File: spec/controllers/tournaments_controller_spec.rb (hunk 13 above)
Comment (Malaber) [reply spec/controllers/tournaments_controller_spec.rb]: accident; readded

File: spec/controllers/tournaments_controller_spec.rb (hunk 12 above)
Comment (Malaber) [reply spec/controllers/tournaments_controller_spec.rb]: removed

File: spec/controllers/tournaments_controller_spec.rb (hunk 10 above)
Comment (Malaber) [reply spec/controllers/tournaments_controller_spec.rb]: done

**Review by Thor77 — DISMISSED** [review]

**Review by Thor77 — APPROVED** [review]
